datasimulator/graph.py

```diff
@@ -0,0 +1,265 @@
+            if skip:
+                logger.error(msg)
+            else:
+                raise DictionaryError(message=msg)
+
+        node.required_links.append(
+            {"node": node_parent, "multiplicity": multiplicity, "name": link_name}
+        )
+
+    def graph_validation(self, required_only=False):
+        """
+        Call to all node validation to validate
+        """
+        return all(
+            [
+                node.node_validation(required_only=required_only)[0]
+                for node in self.nodes
+            ]
+        )
+
+    def construct_graph_edges(self):
+        """
+        Construct edges between nodes. Ignore option links
+        """
+
+        # Link nodes together to create graph
+        for node in self.nodes:
+            if node == self.root:
+                continue
+            if not node.links:
+                logger.error(
+                    "ERROR: {} should have at least one link to other node".format(
+                        node.name
+                    )
+                )
+            try:
+                node_links = node.links
+
+                if not isinstance(node_links, list):
+                    node_links = [node_links]
+
+                # expect node_links contains list of links
+                for link in node_links:
+                    if isinstance(link, dict):
+                        if not link.get("required"):
+                            continue
+                        if "target_type" in link:
+                            self._add_required_link_to_node(
+                                node,
+                                link["target_type"],
+                                link.get("name"),
+                                link.get("multiplicity"),
+                            )
+
+                        if "sub_group" in link or "subgroup" in link:
+                            sub_links = link.get("sub_group") or link.get("subgroup")
+                            if not isinstance(sub_links, list):
+                                sub_links = [sub_links]
+
+                            # just pick one of sub-group links
+                            for sub_link in sub_links:
+                                if "target_type" in sub_link:
+                                    self._add_required_link_to_node(
+                                        node,
+                                        sub_link["target_type"],
+                                        sub_link.get("name"),
+                                        sub_link.get("multiplicity"),
+                                    )
+                                    break
+
+            except TypeError as e:
+                raise DictionaryError(
+                    "Node {} have non-list links. Detail {}".format(
+                        node.name, e.message
+                    )
+                )
+
+    def generate_submission_order_path_to_node(self, node, submission_order):
```

Comment:
make this a closure inside `generate_submission_order`

Reply:
This one function is helpful for integration test latter. We want have a test to submit current node.

Reply:
in that case I would refactor this to not apply side effects to `submission_order` (not take it as an argument) and just return a new list, so it will be easier to use where it gets called

Reply:
Agree non-recursive is better than recursive. I refactor it in e6bd4c559e8de5f1b2fe8b05faa47c8103c54e6f